### PR TITLE
Search projects in parallel.  Cancel any parallel work once enough results are found.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.cs [hunk 1]
```diff
@@ -228,40 +236,60 @@ internal abstract partial class AbstractAddImportCodeFixProvider<TSimpleNameSynt
...
 
-                // Add its results to the final result set we're keeping.
-                AddRange(allSymbolReferences, await doneTask.ConfigureAwait(false));
+                    // Add its results to the final result set we're keeping.
+                    AddRange(allSymbolReferences, await doneTask.ConfigureAwait(false));
 
-                // If we've got enough, no need to keep searching. 
-                // Note: We do not cancel the existing tasks that are still executing.  These tasks will
-                // cause our indices to be created if necessary.  And that's good for future searches which
-                // we will invariably perform.
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.cs]: This comment was no longer valid.  Searching from Add-Import did not cause indices to be created.  Indices are created independently by the special incremental analyzer we have for that purpose.

So these extra tasks were literally wasting CPU.  Now we at least try to kill them off once we find enough project or metadata results.

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

File: src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.cs [hunk 2]
```diff
@@ -228,40 +236,60 @@ internal abstract partial class AbstractAddImportCodeFixProvider<TSimpleNameSynt
...
+
+                await WaitForTasksAsync(allSymbolReferences, findTasks, nestedTokenSource, cancellationToken).ConfigureAwait(false);
             }
+        }
 
-            while (findTasks.Count > 0)
+        private async Task WaitForTasksAsync(
+            List<SymbolReference> allSymbolReferences,
+            HashSet<Task<List<SymbolReference>>> findTasks,
+            CancellationTokenSource nestedTokenSource,
```
Comment (dpoeschl) [inline src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.cs]: nestedWorkCancellationToken? parallelWorkCancellationToken? cancellationTokenSourceForNested(/Parallel)Work?

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.cs]: nestedWorkCancellationTokenSource sounds fine to me.

**Comment by CyrusNajmabadi** [comment]
retest  prtest/win/dbg/eta  please

**Comment by CyrusNajmabadi** [comment]
retest prtest/win/vsi/p0 please

**Comment by CyrusNajmabadi** [comment]
retest prtest/win/vsi/p1 please

**Comment by dpoeschl** [comment]
Can we test this? Do we have tests for this? How are we dealing with the nondeterminism?

**Comment by CyrusNajmabadi** [comment]
> Can we test this? 

All our P2P tests are testing this codepath already.  I can try to add some more if you'd like.  Potentially some simulating solutions with many projects in them. 

> How are we dealing with the nondeterminism?

The same way any test would deal with non-determinism.  i.e. if we would have more results than the cutoff amount, then the test would simply assert: "I should get at exactly 3 results, and those 3 should be from this set of N possible results".  i.e. they test exactly what we're promising to find for the user.  Some set of matching results that we think are good enough.

**Comment by dpoeschl** [comment]
> All our P2P tests are testing this codepath already.

Can you point me to what you're referring to?

Assuming those exist and cover this, :+1:

**Comment by CyrusNajmabadi** [comment]
Need Senior sign off.  THanks!

**Comment by Pilchie** [comment]
This looks okay to me, but are you sure you're actually cancelling often enough?  I only see calls to `ThrowIfCancellationRequested` at the entrance to methods, none inside any loops that might actually take any time ...

:+1:, but might need followup.

**Comment by CyrusNajmabadi** [comment]
I'll do a pass to see if there are more places to check.  importantly though, we are at least attempting to cancel all the tasks we spawn off once we get enough results.  this can't be any worse than *not* cancelling as we are functioning today.

**Pull request merged**